File: NSGA2.cpp
#include "NSGA2.h"
#include <algorithm>

struct less_than_key
{
    inline bool operator() (const RouteSet& struct1, const RouteSet& struct2)
    {
        return (struct1.crowding_distance > struct2.crowding_distance);
    }
};

NSGA2::NSGA2(int generations, int num_of_individuals, TNDP tndp, int num_of_routes, int num_of_tour_particips, float tournament_prob)
{
    this->generations = generations;
    this->num_of_individuals = num_of_individuals;
    this->network_graph = tndp.network_graph;
    this->demand_matrix = tndp.demand_matrix;
    this->network_size = tndp.size;
    this->population = Population();
    this->num_of_routes = num_of_routes;
    this->num_of_tour_particips = num_of_tour_particips;
    this->tournament_prob = tournament_prob;
}
void NSGA2::explore(std::vector<bool> & visited, int node, int node_from, Route route, int max_length)
{
    if (!visited[node])
    {
        std::vector<int> cand_nodes;
        for (const auto x : this->network_graph.nodes[node])
        {
            if (x.to != node_from && !visited[x.to])
            {
                cand_nodes.push_back(x.to);
            }
        }
        if (cand_nodes.size() == 0)
        {
            return;
        }

        int neighbour = rand() % cand_nodes.size();
        route.nodes.push_back(cand_nodes[neighbour]);

        if (route.size < max_length)
        {
            this->explore(visited, cand_nodes[neighbour], node, route, max_length);
        }
    }
}
Route NSGA2::generate_random_route()
{
    int random_start_point = rand() % this->network_size;
    int max_length = rand() % 10 + 4;
    std::vector<bool> visited(this->network_size, false);
    Route random_route;
    random_route.nodes.push_back(random_start_point);
    this->explore(visited, random_start_point, -1, random_route, max_length);
}
RouteSet NSGA2::generate_individual()
{
    RouteSet random_routeset = RouteSet(this->network_graph, this->demand_matrix);
    for (int i = 0; i < this->num_of_routes; i++)
    {
        Route random_route = this->generate_random_route();
        random_routeset.routes.push_back(random_route);
    }
    return random_routeset;
}

Population NSGA2::initialize_population()
{
    Population population;
    for (int i = 0; i < this->num_of_individuals; i++)
    {
        RouteSet individual = this->generate_individual();
        individual.calculate_objectives();
        population.all_population.push_back(individual);
    }
    return population;
}

void NSGA2::fast_non_dominated_sort(Population & population)
{
    this->population.fronts = {{}};
    for (auto & individual: population.all_population)
    {
        individual.domination_count = 0;
        individual.dominated_solutions = {};
        for (auto & other_individual : population.all_population)
        {
            if (individual.dominates(other_individual))
            {
                individual.dominated_solutions.push_back(other_individual);
            }
            else if (other_individual.dominates(individual))
            {
                individual.domination_count++;
            }
        }
        if (individual.domination_count == 0)
        {
            individual.rank = 0;
            population.fronts[0].push_back(individual);
        }
    }
    int i = 0;
    while (population.fronts[i].size() > 0)
    {
        std::vector<RouteSet> temp;
        for (auto & individual : population.fronts[i])
        {
            for (auto & other_individual : individual.dominated_solutions)
            {
                other_individual.domination_count--;
                if (other_individual.domination_count == 0)
                {
                    other_individual.rank = i + 1;
                    temp.push_back(other_individual);
                }
            }
        }
        i++;
        population.fronts.push_back(temp);
    }
}

void NSGA2::calculate_crowding_distance(std::vector<RouteSet> & front)
{
    if (front.size() > 0)
    {
        int solutions_num = front.size();
        for (auto & individual : front)
        {
            individual.crowding_distance = 0;
        }

        for (int m = 0; m < front[0].objectives.size(); m++)
        {
            front[0].crowding_distance = INT32_MAX;
            front[solutions_num - 1].crowding_distance = INT32_MAX;
            std::vector<double> m_values;
            for (const auto & individual : front)
            {
                m_values.push_back(individual.objectives[m]);
            }
            auto it_max = max_element(std::begin(m_values), std::end(m_values));
            auto it_min = min_element(std::begin(m_values), std::end(m_values));

            int scale = *it_max - *it_min;

            if (scale == 0) scale = 1;

            for (int i = 1; i < solutions_num - 1; i++)
            {
                front[i].crowding_distance += (front[i + 1].objectives[m] - front[i - 1].objectives[m]) / scale;
            }
        }
    }
}

int NSGA2::crowding_operator(const RouteSet & individual, const RouteSet & other_individual)
{
    if ((individual.rank < other_individual.rank) || ((individual.rank == other_individual.rank) &&
     individual.crowding_distance > other_individual.crowding_distance))
    {
        return 1;
    }
    return -1;
}

std::vector<RouteSet> NSGA2::create_children(Population population)
{
    std::vector<RouteSet> children;
    while (children.size() < population.all_population.size())
    {
        RouteSet parent1 = this->tournament(population);
        RouteSet parent2 = this->tournament(population);

        while (parent1.routes == parent2.routes)
        {
            parent2 = this->tournament(population);
        }
        std::vector<RouteSet> offspring = this->crossover(parent1, parent2);
        RouteSet child1 = offspring[0];
        RouteSet child2 = offspring[1];
        this->mutate(child1);
        this->mutate(child2);
        child1.calculate_objectives();
        child2.calculate_objectives();

        children.push_back(child1);
        children.push_back(child2);
    }
    return children;
}


RouteSet NSGA2::tournament(Population population)
{
    std::vector<RouteSet> participants;
    std::vector<bool> taken(population.all_population.size(), false);
    
    int n = 0;
    while (n < num_of_tour_particips)
    {
        int participant = rand() % population.all_population.size();
        while (taken[participant])
        {
            participant = rand() % population.all_population.size();
        }
        taken[participant] = true;
    }
    RouteSet * best = nullptr;
    for (auto & participant : participants)
    {
        if (best == nullptr || (this->crowding_operator(participant, *best) == 1 && this->choose_with_prob(this->tournament_prob)))
        {
            best = &participant;
        }
    }

    return *best;

}

bool NSGA2::choose_with_prob(float prob)
{
    int n = rand() % 100000;
    float p = (float) n / 100000.f;
    return p <= prob;
}

std::vector<RouteSet> NSGA2::crossover(RouteSet parent1, RouteSet parent2)
{   
    RouteSet child1 = parent1;
    RouteSet child2 = parent2;

    int longest_route_ind1 = child1.routes[0].size;
    int pos_longest_route_ind1 = 0;

    for (int i = 0; i < child1.routes.size(); i++)
    {
        if (child1.routes[i].nodes.size() > longest_route_ind1)
        {
            longest_route_ind1 = child1.routes[i].nodes.size();
            pos_longest_route_ind1 = 1;
        }
    }
    Route parent1_longest_route = child1.routes[pos_longest_route_ind1];
    child1.routes.erase(child1.routes.begin() + pos_longest_route_ind1);

    
    int longest_route_ind2 = child2.routes[0].size;
    int pos_longest_route_ind2 = 0;

    for (int i = 0; i < child2.routes.size(); i++)
    {
        if (child2.routes[i].nodes.size() > longest_route_ind2)
        {
            longest_route_ind2 = child2.routes[i].nodes.size();
            pos_longest_route_ind2 = 1;
        }
    }
    Route parent2_longest_route = child2.routes[pos_longest_route_ind2];
    child2.routes.erase(child2.routes.begin() + pos_longest_route_ind2);

    child1.routes.push_back(parent2_longest_route);
    child2.routes.push_back(parent1_longest_route);

    return std::vector<RouteSet>{child1, child2};
}

void NSGA2::mutate(RouteSet routeset)
{
    for (const auto & node : routeset.routes[0].nodes)
    {
        for (int i = 1; i < routeset.routes.size(); i++)
        {
            int end_point = routeset.routes[i].nodes.size();
            if (node == routeset.routes[i].nodes[0])
            {
                if (routeset.routes[i].nodes.size() > 2)
                {
                    routeset.routes[i].nodes.erase(routeset.routes[i].nodes.begin());
                }
            }
            else if (node == routeset.routes[i].nodes[end_point])
            {
                if (routeset.routes[i].nodes.size() > 2)
                {
                    routeset.routes[i].nodes.erase(routeset.routes[i].nodes.begin() + end_point);
                }
            }
        }
    }
}

std::vector<RouteSet> NSGA2::run()
{
    this->population = this->initialize_population();
    this->fast_non_dominated_sort(this->population);

    for (auto & front : this->population.fronts)
    {
        this->calculate_crowding_distance(front);
    }    
    std::vector<RouteSet> children = this->create_children(this->population);
    Population * returned_population = nullptr; 

    for (int g = 0; g < this->generations; g++)
    {   
        this->population.all_population.insert(this->population.all_population.end(), children.begin(), children.end());
        this->fast_non_dominated_sort(this->population);
        Population new_population;
        int front_num = 0;

        while (new_population.all_population.size() + this->population.fronts[front_num].size() <= this->num_of_individuals)
        {
            this->calculate_crowding_distance(this->population.fronts[front_num]);
            for (auto & rs: this->population.fronts[front_num])
            {
                new_population.all_population.push_back(rs);
            }
            front_num++;
        }
        this->calculate_crowding_distance(this->population.fronts[front_num]);
        std::sort(this->population.fronts[front_num].begin(), this->population.fronts[front_num].end(), less_than_key());
        new_population.all_population.insert(new_population.all_population.end(), this->population.fronts[front_num].begin(), this->population.fronts[front_num].end());
        returned_population = &this->population;
        this->population = new_population;
        this->fast_non_dominated_sort(this->population);
        for (auto front : this->population.fronts)
        {
            this->calculate_crowding_distance(front);
        }
        children = this->create_children(this->population);
    }
    return returned_population->fronts[0]; 
}  

File: nsga2.py
import random
import numpy as np
from population import Population
from routeset import RouteSet
import copy

class NSGAII:
    def __init__(self, generations, num_of_individuals, tndp, num_of_routes, num_of_tour_particips, tournament_prob):
        self.generations = generations
        self.num_of_individuals = num_of_individuals
        self.graph = tndp.network
        self.demand_matrix = tndp.demand_matrix
        self.population = None
        self.num_of_routes = num_of_routes
        self.network_size = tndp.size
        self.num_of_tour_particips = num_of_tour_particips
        self.tournament_prob = tournament_prob


    def explore(self, visited, node, node_from, route, max_length):
        if node not in visited:
            cand_nodes = []
            for x in self.graph.nodes[node]:
                if x.to != node_from and x.to not in visited:
                    cand_nodes.append(x.to)

            if len(cand_nodes) == 0:
                return
            
            neighbour = random.randrange(0, len(cand_nodes))
            route.append(cand_nodes[neighbour])

            if len(route) < max_length:
                self.explore(visited, cand_nodes[neighbour], node, route, max_length)

    def generate_random_route(self):
        random_start_point = random.randrange(0, self.network_size)
        max_length = random.randrange(11, 25) + 1
        visited = []
        random_route = [random_start_point]
        self.explore(visited, random_start_point, -1, random_route, max_length)
        return random_route
    
    def generate_individual(self):
        random_routeset = RouteSet()
        for i in range(self.num_of_routes):
            random_route = self.generate_random_route()
            random_routeset.routes.append(random_route)
        return random_routeset
            
    def initialize_population(self):
        population = Population()
        for _ in range(self.num_of_individuals):
            individual = self.generate_individual()
            individual.calculate_objectives(self.graph, self.demand_matrix)
            population.append(individual)
        return population

    def fast_non_dominated_sort(self, population):
        population.fronts = [[]]
        for individual in population:
            individual.domination_count = 0
            individual.dominated_solutions = []
            for other_individual in population:
                if individual.dominates(other_individual):
                    individual.dominated_solutions.append(other_individual)
                elif other_individual.dominates(individual):
                    individual.domination_count += 1
            if individual.domination_count == 0:
                individual.rank = 0
                population.fronts[0].append(individual)
        i = 0
        while len(population.fronts[i]) > 0:
            temp = []
            for individual in population.fronts[i]:
                for other_individual in individual.dominated_solutions:
                    other_individual.domination_count -= 1
                    if other_individual.domination_count == 0:
                        other_individual.rank = i + 1
                        temp.append(other_individual)
            i = i + 1
            population.fronts.append(temp)

    
    def calculate_crowding_distance(self, front):
        if len(front) > 0:
            print(len(front))
            solutions_num = len(front)
            for individual in front:
                individual.crowding_distance = 0

            for m in range(len(front[0].objectives)):
                front.sort(key=lambda individual: individual.objectives[m])
                front[0].crowding_distance = 10 ** 9
                front[solutions_num - 1].crowding_distance = 10 ** 9
                m_values = [individual.objectives[m] for individual in front]
                scale = max(m_values) - min(m_values)
                if scale == 0: scale = 1
                for i in range(1, solutions_num - 1):
                    front[i].crowding_distance += (front[i + 1].objectives[m] - front[i - 1].objectives[m]) / scale
            print(len(front))
            
    
    def crowding_operator(self, individual, other_individual):
        if (individual.rank < other_individual.rank) or \
                ((individual.rank == other_individual.rank) and (
                        individual.crowding_distance > other_individual.crowding_distance)):
            return 1
        else:
            return -1
    def create_children(self, population):
        children = []
        while len(children) < len(population):
            parent1 = self.tournament(population)
            parent2 = self.tournament(population)

            while parent1.routes == parent2.routes:
                parent2 = self.tournament(population)
            
            child1, child2 = self.crossover(parent1, parent2)
            self.mutate(child1)
            self.mutate(child2)
            child1.calculate_objectives(self.graph, self.demand_matrix)
            child2.calculate_objectives(self.graph, self.demand_matrix)
            children.append(child1)
            children.append(child2)

        return children

    def tournament(self, population):
        participants = random.sample(population.population, self.num_of_tour_particips)
        best = None
        for participant in participants:
            if best is None or (
                    self.crowding_operator(participant, best) == 1 and self.choose_with_prob(self.tournament_prob)):
                best = participant

        return best

    def choose_with_prob(self, prob):
        if random.random() <= prob:
            return True
        

    def crossover(self, parent1, parent2):
        child1 = copy.deepcopy(parent1)
        child2 = copy.deepcopy(parent2)

        longest_route_ind1 = len(child1.routes[0])
        pos_longest_route_ind1 = 0
        for i in range(len(child1.routes)):
            if len(child1.routes[i]) > longest_route_ind1:
                longest_route_ind1 = len(child1.routes[i])
                pos_longest_route_ind1 = i

        parent1_longest_route = child1.routes[pos_longest_route_ind1]
        child1.routes.pop(pos_longest_route_ind1)

        longest_route_ind2 = len(child2.routes[0])
        pos_longest_route_ind2 = 0
        for i in range(len(child2.routes)):
            if len(child2.routes[i]) > longest_route_ind2:
                longest_route_ind2 = len(child2.routes[i])
                pos_longest_route_ind2 = i


        parent2_longest_route = child2.routes[pos_longest_route_ind2]

        child2.routes.pop(pos_longest_route_ind2)

        child1.routes.append(parent2_longest_route)
        child2.routes.append(parent1_longest_route)
        return child1, child2

    def mutate(self, route_set):
        for node in route_set.routes[0]:   
            for i in range(1,len(route_set.routes)):
                end_point = len(route_set.routes[i]) - 1
                if node == route_set.routes[i][0]:
                    if len(route_set.routes[i]) > 2:
                        route_set.routes[i].pop(0)
                elif node == route_set.routes[i][end_point]:
                    if len(route_set.routes[i]) > 2:
                        route_set.routes[i].pop(end_point)



    def run(self):
        self.population = self.initialize_population()

        self.fast_non_dominated_sort(self.population)
        for front in self.population.fronts:
            self.calculate_crowding_distance(front)
        children = self.create_children(self.population)
        returned_population = None


        for _ in range(self.generations):
            print("Generacion {}".format(_))
            self.population.extend(children)
            self.fast_non_dominated_sort(self.population)
            new_population = Population()
            front_num = 0
            print(len(self.population.fronts))
            while len(new_population) + len(self.population.fronts[front_num]) <= self.num_of_individuals:
                self.calculate_crowding_distance(self.population.fronts[front_num])
                for rs in self.population.fronts[front_num]:
                    new_population.append(copy.deepcopy(rs))
                front_num += 1

            self.calculate_crowding_distance(self.population.fronts[front_num])
            self.population.fronts[front_num].sort(key=lambda individual: individual.crowding_distance, reverse=True)
            new_population.extend(self.population.fronts[front_num][0:self.num_of_individuals - len(new_population)])
            returned_population = self.population
            self.population = new_population
            self.fast_non_dominated_sort(self.population)
            for front in self.population.fronts:
                self.calculate_crowding_distance(front)
            children = self.create_children(self.population)

            for routeset in returned_population.fronts[0]:
                print("part:")
                for route in routeset.routes:
                    for a in route:
                        print("{} - ".format(a), end="")
                    print("")
        return returned_population.fronts[0]
            
